src/Strategy/Maven/DepTree.hs

```diff
@@ -0,0 +1,174 @@
+module Strategy.Maven.DepTree (
+  analyze,
+  parseDotGraph,
+
+  -- * Exported for testing
+  DotGraph (..),
+  PackageId (..),
+) where
+
+import Control.Algebra (Has, run)
+import Control.Applicative ((<|>))
+import Control.Carrier.State.Strict (execState, modify)
+import Control.Effect.Diagnostics (Diagnostics, context)
+import Control.Effect.Lift (Lift, sendIO)
+import Data.Char (isSpace)
+import Data.Foldable (for_, traverse_)
+import Data.String (IsString)
+import Data.String.Conversion (toString, toText)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Void (Void)
+import DepTypes (
+  DepEnvironment (..),
+  DepType (MavenType),
+  Dependency (..),
+  VerConstraint (CEq),
+ )
+import Discovery.Walk (
+  WalkStep (WalkContinue),
+  findFileNamed,
+  walk,
+ )
+import Effect.Exec (AllowErr (Never), Command (..), Exec, execThrow)
+import Effect.Grapher (direct, edge, evalGrapher)
+import Effect.ReadFS (ReadFS, readContentsParser)
+import Graphing (Graphing, gmap)
+import Path (Abs, Dir, File, Path)
+import Path.IO (removeFile)
+import System.Random (randomIO)
+import Text.Megaparsec (
+  Parsec,
+  between,
+  many,
+  takeWhile1P,
+  try,
+ )
+import Text.Megaparsec.Char (space1)
+import Text.Megaparsec.Char.Lexer qualified as Lexer
+import Types (GraphBreadth (Complete))
+
+newtype DepTreeLabel
+  = BuildTag DepEnvironment
+  deriving (Eq, Ord, Show)
+
+outputFileName :: IsString a => a
+outputFileName = "fossa-deptree.dot"
+
+analyze :: (Has Exec sig m, Has ReadFS sig m, Has Diagnostics sig m, Has (Lift IO) sig m) => Path Abs Dir -> m (Graphing Dependency, GraphBreadth)
+analyze dir = do
+  -- We generate files with a random prefix so we don't pick up files from previous runs.
+  -- If we did pick them up, they're likely not enough of a conflict to worry about, but
+  -- it's better than not doing it.
+  randIdent <- abs <$> sendIO randomIO
+  _ <- context "Running maven 'dependency:tree' plugin" $ execThrow dir $ deptreeCmd randIdent
+  graphFiles <- context "Locating maven output files" $ findDepTreeOutputs dir randIdent
+  graphs <- context "Parsing output files" $ traverse (readContentsParser parseDotGraph) graphFiles
+  -- TODO: Will this work correctly in the face of exceptions?
```

Comment:
an exception here will abort analysis / later files in the `graphFiles` list wouldn't be deleted

Reply:
What's the right thing to do here to make sure the files get cleaned up? `catch`?

Reply:
something like
```hs
import Control.Exception.Extra (safeCatch)

ignoreExceptions :: Has (Lift IO) sig m => m a -> m ()
ignoreExceptions act = void act `safeCatch` (\(_ :: SomeException) -> pure ())
```

`traverse (ignoreExceptions . removeFile) ...`

(maybe even adding this to `Control.Exception.Extra`)

Reply:
Ah, I see. I think I'm asking a different question - what's the right way to handle exceptions in L66 (i.e. in `readContentsParser`)?

In particular, I know `ReadFS` is an effect - is there a specific catching effect that I should use?

Reply:
ReadFS catches IO exceptions and reports them as diagnostics.  [See here](https://github.com/fossas/spectrometer/blob/72d79a21505522c84478a86a811f76eb9db59e44/src/Effect/ReadFS.hs#L199-L218) for the code that does this.

Reply:
If `readContentsParser` fails, it'll abort with a diagnostics error as wes mentions. To ensure cleaning up files, you can use the suggestion above + chain `finally` off of line 66:

```hs
import Control.Effect.Exception (finally)

[...]
graphs <- (context "..." $ ... ) `finally` deleteAllTheFiles
[...]
```

or maybe create a helper for this too
```hs
withCleanup :: Has (Lift IO) sig m -> m b -> m a -> m a
withCleanup = flip finally
```

```hs
[...]
graphFiles <- ...
withCleanup (deleteAllThe graphFiles) $ do
  graphs <- ...
  pure (buildGraph ...)